### DISCUSSION TITLE
HammerDB on remote windows server vs on DB server

### DISCUSSION THREAD

**Discussion by finnyvargheese** [TPC-Council/HammerDB#649] [General]
Need some help understanding the below results that are captured during hammerDB tetsing,
I have installed hammerDB on a windows node(Remote) and running hammerdb against oracle database (DB server).

**HammerDB 4.8**
Remote server: Installed on windows server with 16 GB ram and 8 core cpu
Database Server: oracle 19c Linux with 10 core cpu 256 GB RAM
Warehouses: 4000
Ramup time: 5 mins
Test time: 15 mins
**200 vusers**
"TEST RESULT : System achieved 185502 NOPM from 322642 Oracle TPM"( test completed within 30 mins)
**250 vusers**
"TEST RESULT : System achieved 1106849 NOPM from 304149 Oracle TPM" ( test took 2 hours to complete, observed ~20 GB increase in the database size(dba_data_files table))
**300 vusers**
Test was running for more than 8 hours and we found increase in the database size about 450 GB during the test. We had to abort the test.

During the 250/300 vusers test, we observed database locks, it got cleared after hours for 250 users test. But the other one kept running and could see the database lock count and the database size was increasing
select
   s.sid,
   s.serial#,
   s.machine,
   lo.oracle_username,
   lo.os_user_name,
   ao.object_name,
   lo.locked_mode
from
   v$locked_object lo, dba_objects ao, v$session s
where
   ao.object_id = lo.object_id and lo.session_id = s.sid;

SELECT sum(bytes)/1024/1024 AS "Database Size (MB)" FROM dba_data_files;

1. Is this normal behaviour ? and Any inputs on analyzing the results ?
2. Do we need to consider equal CPU cores for remote windows server as DB server when we perform full load test? 
3. Do we need to consider equal number of virtual users as CPU threads?
4. Does it vary the hammerDB results with remote server?

**Comment by sm-shaw** [comment]
Notwithstanding that the data for the 250Vuser test looks incorrect in comparison, You are making one of the most common errors in database benchmarking by over-testing.  what you should do is generate a [performance profile](https://www.hammerdb.com/blog/uncategorized/hammerdb-using-mysql-5-7-vs-8-0-to-understand-performance-profiles/). 

For example, below is a test on a vendor provided binary and a compiled from source version of the same software. 

So we test by taking multiple points on the curve. In this case, we have shown that the performance is very similar and peaks at 80 Vusers. 

<img width="399" alt="profile" src="https://github.com/TPC-Council/HammerDB/assets/38044085/8a80ac59-997f-4059-b6bd-b470beb82acc">


In your case, you have started testing at 200, 250 and 300 Vusers - whereas the maximum performance at 10 cores is likely to be much lower. 

With every database benchmarking scenario the advice is always to start testing with 1 Virtual User and then build up in increments. In the GUI there is the autopilot feature and in the CLI you can use scripts to do this. 

Do this and you will not have the issues you have encountered by over-configuring the test for the system you have.

**Reply by finnyvargheese** [reply]
Thanks @sm-shaw for the detailed explanation.
